### DISCUSSION TITLE
Confusion about unplugged packages

### DISCUSSION THREAD

**Discussion by UNLRN** [yarnpkg/berry#3197] [Q&A] [answered]
I've been running a repo with zero-installs and have run into a couple issues that have been solved through looking at issues here. One thing that keeps popping up is the scenario of 'unplugged' packages. The docs say that that:

> Packages are unplugged in two cases: either explicitly when the unplugged field is set to true, or implicitly when the package lists postinstall scripts.

And [this issue](https://github.com/yarnpkg/berry/issues/1567) states that most `postinstall` scripts are actually okay to disable, anyways. So when I see a notification that a package has build scripts, I'll often go double check to make sure they aren't _actually_ needed.

All that makes sense. So I've set `enableScripts: false` to disable build scripts, which I would assume would prevent packages from becoming unplugged unless explicitly stated. However, I am still getting some packages unplugged. Does this scenario only apply to the direct dependencies `postinstall` script? For example, if a package I `yarn add` has a dependency on another package that has a `postinstall` script, will that package become unplugged? It seems like that' is what I'm noticing happening. (the [package](https://github.com/DataDog/dd-trace-js) in question as a reference)

Is there any way to move forward with a true `zero-install` repo when a package becomes unplugged if its a dependency of the package? Or will we always have to run the `yarn install` step so that those individual packages will get added to the target machine?

**Comment by merceyz** [comment, marked as answer]
There is a third case where packages are unplugged that isn't listed in the docs (will fix that), when a package contains native files Yarn will automatically unplug it as well.
https://github.com/yarnpkg/berry/blob/344818b6da13b3774231569f02f1476f606d9a4c/packages/plugin-pnp/sources/jsInstallUtils.ts#L65-L74

**Reply by arcanis** [reply]
Also note that you can always override that using the `dependenciesMeta.unplugged` field

**Reply by UNLRN** [reply]
@merceyz Thanks for the additional scenario, that could explain why I'm still seeing some packages get unplugged.

@arcanis Yeah, I saw that field in the docs but I was unsure when to use it. Is there something in particular to look for when overriding the `unplugged` logic? Or is pretty guaranteed that when the third case is hit, it's required for the package to be unplugged? I'd really like to get to `zero-installs` but understand if there is a limitation with certain packages needing to be built on the target machines

**Reply by arcanis** [reply]
It really depends on the package. Some (`node-notifier`) include a binary within their sources that they need to run properly, in which case unplugging them is the only option. Others have fallbacks on wasm, or include their C/C++ sources for completeness, and in that case unplugging them isn't necessary. Usually it's the former, though.

**Reply by UNLRN** [reply]
@arcanis Ah, so in the scenarios where you have a package that _has_ to get unplugged, `zero-install` becomes `some-installs`? Meaning you can still use the `zero-install` approach, however you'll have to run `yarn install` for those packages to get unplugged on target machines (since they aren't included in source and will have to get unzipped in order to function correctly?)

**Reply by arcanis** [reply]
Yep, exactly!

**Reply by minlite** [reply]
@arcanis Would be nice to add this to the docs under [Zero-Installs](https://yarnpkg.com/features/caching#zero-installs). There are many packages such as `node-gyp` (62m+ downloads per week) and `bcrypt` (1.5m+ downloads per week) that are incredibly popular and get unplugged. They are not rare.
